File: apps/demo-be-clojure-pedestal/src/demo_be_cjpd/handlers/test_api.clj
(ns demo-be-cjpd.handlers.test-api
  "Test-only API handlers gated by ENABLE_TEST_API=true env var.
   These endpoints are used by FE E2E tests to reset state between runs."
  (:require [cheshire.core :as json]
            [next.jdbc :as jdbc]
            [demo-be-cjpd.db.protocols :as proto]))

(defn- json-response [status body]
  {:status  status
   :headers {"Content-Type" "application/json"}
   :body    (json/generate-string body)})

(defn- error-response [status message]
  {:status  status
   :headers {"Content-Type" "application/json"}
   :body    (json/generate-string {:message message})})

(defn reset-db-handler!
  "POST /api/v1/test/reset-db — Delete all data in dependency order.
   Deletion order: attachments → expenses → revoked_tokens → users.
   Takes a raw datasource for direct truncation (test infra only)."
  [ds]
  (fn [_request]
    (jdbc/execute! ds ["DELETE FROM attachments"])
    (jdbc/execute! ds ["DELETE FROM expenses"])
    (jdbc/execute! ds ["DELETE FROM revoked_tokens"])
    (jdbc/execute! ds ["DELETE FROM users"])
    (json-response 200 {:message "Database reset successful"})))

(defn promote-admin-handler!
  "POST /api/v1/test/promote-admin — Set a user's role to ADMIN by username.
   Body: {\"username\": \"...\"}
   Returns 200 with message or 404 if user not found."
  [ds user-repo]
  (fn [request]
    (let [params   (or (:json-params request) {})
          username (:username params)]
      (if-not username
        (error-response 400 "Missing required field: username")
        (let [user (proto/find-user-by-username user-repo username)]
          (if-not user
            (error-response 404 (str "User not found: " username))
            (do
              (jdbc/execute! ds
                             ["UPDATE users SET role = 'ADMIN' WHERE username = ?"
                              username])
              (json-response 200 {:message (str "User " username " promoted to ADMIN")}))))))))

File: apps/a-demo-be-ts-effect/src/routes/test-api.ts
import { HttpRouter, HttpServerResponse, HttpServerRequest } from "@effect/platform";
import { Effect } from "effect";
import { SqlClient } from "@effect/sql";
import { UserRepository } from "../infrastructure/db/user-repo.js";
import { NotFoundError } from "../domain/errors.js";

const resetDb = Effect.gen(function* () {
  const sql = yield* SqlClient.SqlClient;
  yield* sql`DELETE FROM attachments`;
  yield* sql`DELETE FROM expenses`;
  yield* sql`DELETE FROM revoked_tokens`;
  yield* sql`DELETE FROM users`;
  return yield* HttpServerResponse.json({ message: "Database reset successful" }, { status: 200 });
});

const promoteAdmin = HttpServerRequest.HttpServerRequest.pipe(
  Effect.flatMap((req) =>
    Effect.gen(function* () {
      const body = yield* req.json as Effect.Effect<Record<string, unknown>, unknown>;
      const username = (body["username"] as string | undefined) ?? "";

      const userRepo = yield* UserRepository;
      const user = yield* userRepo.findByUsername(username);
      if (!user) {
        return yield* Effect.fail(new NotFoundError({ resource: "User" }));
      }

      const sql = yield* SqlClient.SqlClient;
      const now = new Date().toISOString();
      yield* sql`UPDATE users SET role = 'ADMIN', updated_at = ${now} WHERE id = ${user.id}`;

      return yield* HttpServerResponse.json({ message: `User ${username} promoted to ADMIN` }, { status: 200 });
    }),
  ),
);

export const testApiRouter = HttpRouter.empty.pipe(
  HttpRouter.post("/api/v1/test/reset-db", resetDb),
  HttpRouter.post("/api/v1/test/promote-admin", promoteAdmin),
);
